--- orig_pciutils-3.9.0/ls-ecaps.c      2022-11-20 20:06:14.000000000 +0800
+++ pciutils-3.9.0/ls-ecaps.c   2022-12-14 11:22:00.053078200 +0800
@@ -653,8 +653,9 @@
       int prevmatched=0;
       int adjcount=0;
       int prevdev=0;
+      int dev=0;
       printf("RCiEP at Device(s):");
-      for (int dev=0; dev < 32; dev++)
+      for (dev=0; dev < 32; dev++)
         {
          if (BITS(bmap, dev, 1))
            {
@@ -810,7 +811,8 @@
     "CPMU registers",
   };

-  for (int i=0; ; i++)
+  int i;
+  for (i=0; ; i++)
     {
       int pos = where + PCI_CXL_RL_BLOCK1_LO + 8*i;
       if (pos + 7 >= where + len)
