Index: busybox-1.33.2/coreutils/date.c
===================================================================
--- busybox-1.33.2.orig/coreutils/date.c
+++ busybox-1.33.2/coreutils/date.c
@@ -330,11 +330,18 @@ int date_main(int argc UNUSED_PARAM, cha
 		if (date_str[0] != '@')
 			tm_time.tm_isdst = -1;
 		ts.tv_sec = validate_tm_time(date_str, &tm_time);
-
+		ts.tv_nsec = 0;
+#if __GLIBC_MINOR__ < 31
 		/* if setting time, set it */
 		if ((opt & OPT_SET) && stime(&ts.tv_sec) < 0) {
 			bb_perror_msg("can't set date");
 		}
+#else
+
+		if ((opt & OPT_SET) && clock_settime(CLOCK_REALTIME, &ts) < 0) {
+			bb_simple_perror_msg("can't set date");
+		}
+#endif
 	}
 
 	/* Display output */
