--- orig_systemd-243/meson.build	2022-12-08 10:21:44.073372758 +0800
+++ systemd-243/meson.build	2022-12-08 10:23:13.672696705 +0800
@@ -37,7 +37,7 @@
 want_ossfuzz = get_option('oss-fuzz')
 want_libfuzzer = get_option('llvm-fuzz')
 want_fuzzbuzz = get_option('fuzzbuzz')
-if want_ossfuzz + want_libfuzzer + want_fuzzbuzz > 1
+if want_ossfuzz and want_libfuzzer and want_fuzzbuzz
         error('only one of oss-fuzz, llvm-fuzz or fuzzbuzz can be specified')
 endif
 
--- orig_systemd-243/src/network/netdev/fou-tunnel.c
+++ systemd-243/src/network/netdev/fou-tunnel.c
@@ -149,7 +149,10 @@ int config_parse_ip_protocol(
                 void *data,
                 void *userdata) {
 
-        uint8_t *protocol = data;
+        uint8_t *ret = data;
+        unsigned protocol;
+        /* linux/fou.h defines the netlink field as one byte, so we need to reject protocols numbers that
+         * don't fit in one byte. */
         int r;
 
         assert(filename);
@@ -158,11 +161,11 @@ int config_parse_ip_protocol(
         assert(rvalue);
         assert(data);
 
-        assert_cc(IPPROTO_MAX-1 <= UINT8_MAX);
-
         r = parse_ip_protocol(rvalue);
-        if (r < 0) {
-                r = safe_atou8(rvalue, protocol);
+        if (r >= 0)
+                protocol = r;
+        else {
+                r = safe_atou(rvalue, &protocol);
                 if (r < 0)
                         log_syntax(unit, LOG_ERR, filename, line, r,
                                    "Failed to parse IP protocol '%s' for Foo over UDP tunnel, "
@@ -170,7 +173,14 @@ int config_parse_ip_protocol(
                 return 0;
         }
 
-        *protocol = r;
+        if (protocol > UINT8_MAX) {
+                log_syntax(unit, LOG_ERR, filename, line, 0,
+                           "IP protocol '%s' for FooOverUDP tunnel out of range, "
+                           "ignoring assignment: %m", rvalue);
+                return 0;
+        }
+
+        *ret = protocol;
         return 0;
 }
