diff -urN orig_iptables-1.8.6/configure iptables-1.8.6/configure
--- orig_iptables-1.8.6/configure	2020-06-03 19:07:27.000000000 +0800
+++ iptables-1.8.6/configure	2020-06-19 10:01:35.521711807 +0800
@@ -822,6 +822,9 @@
 with_kbuild
 with_ksource
 with_xtlibdir
+with_pfxmods
+with_pf4mods
+with_pf6mods
 enable_ipv4
 enable_ipv6
 enable_largefile
@@ -12460,6 +12463,21 @@
   xtlibdir="${libdir}/xtables"
 fi
 
+# Check whether --with-pfxmods was given.
+if test "${with_pfxmods+set}" = set; then :
+  withval=$with_pfxmods; pfxmods="$withval"
+fi
+
+# Check whether --with-pf4mods was given.
+if test "${with_pf4mods+set}" = set; then :
+  withval=$with_pf4mods; pf4mods="$withval"
+fi
+
+# Check whether --with-pf6mods was given.
+if test "${with_pf6mods+set}" = set; then :
+  withval=$with_pf6mods; pf6mods="$withval"
+fi
+
 # Check whether --enable-ipv4 was given.
 if test "${enable_ipv4+set}" = set; then :
   enableval=$enable_ipv4; enable_ipv4="$enableval"
diff -urN orig_iptables-1.8.6/configure.ac iptables-1.8.6/configure.ac
--- orig_iptables-1.8.6/configure.ac	2020-06-03 18:32:01.000000000 +0800
+++ iptables-1.8.6/configure.ac	2020-06-19 10:01:35.521711807 +0800
@@ -33,6 +33,21 @@
 	[Path where to install Xtables extensions [[LIBEXECDIR/xtables]]]),
 	[xtlibdir="$withval"],
 	[xtlibdir="${libdir}/xtables"])
+AC_ARG_WITH([pfxmods],
+	AS_HELP_STRING([--with-pfxmods=EXTENSIONS],
+ 	[Path where to install Xtables extensions [[LIBEXECDIR/xtables]]]),
+	[pfxmods="$withval"],
+	[pfxmods=""])
+AC_ARG_WITH([pf4mods],
+	AS_HELP_STRING([--with-pf4mods=EXTENSIONS],
+ 	[Path where to install Xtables extensions [[LIBEXECDIR/xtables]]]),
+	[pf4mods="$withval"],
+	[pf4mods=""])
+AC_ARG_WITH([pf6mods],
+	AS_HELP_STRING([--with-pf6mods=EXTENSIONS],
+ 	[Path where to install Xtables extensions [[LIBEXECDIR/xtables]]]),
+	[pf6mods="$withval"],
+	[pf6mods=""])
 AC_ARG_ENABLE([ipv4],
 	AS_HELP_STRING([--disable-ipv4], [Do not build iptables]),
 	[enable_ipv4="$enableval"], [enable_ipv4="yes"])
@@ -82,6 +97,14 @@
 )
 LDFLAGS="$saved_LDFLAGS";
 
+pfx_modules="$pfxmods";
+pf4_modules="$pf4mods";
+pf6_modules="$pf6mods";
+
+AC_SUBST([pfx_modules])
+AC_SUBST([pf4_modules])
+AC_SUBST([pf6_modules])
+
 blacklist_modules=""
 blacklist_x_modules=""
 blacklist_b_modules=""
diff -urN orig_iptables-1.8.6/extensions/GNUmakefile.in iptables-1.8.6/extensions/GNUmakefile.in
--- orig_iptables-1.8.6/extensions/GNUmakefile.in	2020-06-03 18:32:01.000000000 +0800
+++ iptables-1.8.6/extensions/GNUmakefile.in	2020-06-19 10:01:35.521711807 +0800
@@ -50,6 +50,9 @@
 pfa_build_mod := $(filter-out @blacklist_modules@ @blacklist_a_modules@,${pfa_build_mod})
 pf4_build_mod := $(filter-out @blacklist_modules@ @blacklist_4_modules@,${pf4_build_mod})
 pf6_build_mod := $(filter-out @blacklist_modules@ @blacklist_6_modules@,${pf6_build_mod})
+pfx_build_mod := $(filter @pfx_modules@,${pfx_build_mod})
+pf4_build_mod := $(filter @pf4_modules@,${pf4_build_mod})
+pf6_build_mod := $(filter @pf6_modules@,${pf6_build_mod})
 pfx_objs      := $(patsubst %,libxt_%.o,${pfx_build_mod})
 pfb_objs      := $(patsubst %,libebt_%.o,${pfb_build_mod})
 pfa_objs      := $(patsubst %,libarpt_%.o,${pfa_build_mod})
diff -urN orig_iptables-1.8.6/iptables/Makefile.am iptables-1.8.6/iptables/Makefile.am
--- orig_iptables-1.8.6/iptables/Makefile.am	2020-06-03 18:32:01.000000000 +0800
+++ iptables-1.8.6/iptables/Makefile.am	2020-06-19 10:01:35.525711768 +0800
@@ -7,7 +7,7 @@
 BUILT_SOURCES =

 common_sources = iptables-xml.c xtables-multi.h xshared.c xshared.h
-common_ldadd   = ../extensions/libext.a ../libxtables/libxtables.la -lm
+common_ldadd   = ../extensions/libext.a ../libxtables/libxtables.la -lm -ldl
 common_cflags  = ${AM_CFLAGS}
 if ENABLE_STATIC
 common_cflags += -DALL_INCLUSIVE
@@ -104,7 +104,6 @@
 
 # Using if..fi avoids an ugly "error (ignored)" message :)
 install-exec-hook:
-	-if test -z "${DESTDIR}"; then /sbin/ldconfig; fi;
 	${INSTALL} -dm0755 "${DESTDIR}${bindir}";
 	for i in ${vx_bin_links}; do ${LN_S} -f "${sbindir}/xtables-legacy-multi" "${DESTDIR}${bindir}/$$i"; done;
 	for i in ${v4_sbin_links}; do ${LN_S} -f xtables-legacy-multi "${DESTDIR}${sbindir}/$$i"; done;
diff -urN orig_iptables-1.8.6/iptables/Makefile.in iptables-1.8.6/iptables/Makefile.in
--- orig_iptables-1.8.6/iptables/Makefile.in	2020-06-03 19:07:28.000000000 +0800
+++ iptables-1.8.6/iptables/Makefile.in	2020-06-19 10:01:35.525711768 +0800
@@ -1550,7 +1550,6 @@
 
 # Using if..fi avoids an ugly "error (ignored)" message :)
 install-exec-hook:
-	-if test -z "${DESTDIR}"; then /sbin/ldconfig; fi;
 	${INSTALL} -dm0755 "${DESTDIR}${bindir}";
 	for i in ${vx_bin_links}; do ${LN_S} -f "${sbindir}/xtables-legacy-multi" "${DESTDIR}${bindir}/$$i"; done;
 	for i in ${v4_sbin_links}; do ${LN_S} -f xtables-legacy-multi "${DESTDIR}${sbindir}/$$i"; done;
