--- orig_htpdate-1.3.7/htpdate.c	2023-01-23 23:54:49.000000000 +0800
+++ htpdate-1.3.7/htpdate.c	2023-02-06 14:07:02.757900607 +0800
@@ -112,9 +112,9 @@ static void printlog(int is_error, char
     va_end(args);
 
     if (logmode)
-        syslog(is_error?LOG_WARNING:LOG_INFO, "%s", buf);
+        syslog(is_error?LOG_WARNING:LOG_INFO, "htpdate: %s", buf);
     else
-        fprintf(is_error?stderr:stdout, "%s\n", buf);
+        fprintf(is_error?stderr:stdout, "htpdate: %s\n", buf);
 }
 
 
@@ -713,6 +713,7 @@ int main(int argc, char *argv[]) {
     int             sleeptime = minsleep;
     int             sw_uid = 0, sw_gid = 0;
     time_t          starttime = 0;
+    int                 initset = 1; /*different set method for init time set*/    
 
     struct passwd   *pw;
     struct group    *gr;
@@ -949,8 +950,16 @@ int main(int argc, char *argv[]) {
 
             /* Do I really need to change the time?  */
             if (sumtimes || !(daemonize || foreground)) {
-                if (setclock(timeavg, setmode) < 0)
-                    printlog(1, "Time change failed");
+                if ( initset && (setmode == 1) && timeavg ) {
+                    printlog( 1, "Initial time set" );
+                    if ( setclock( timeavg, 2 ) == 0 )
+                        initset = 0;
+                }
+                else {
+		    /* Correct the clock, if not in "adjtimex" mode */
+		    if ( setclock( timeavg, setmode ) < 0 )                
+                        printlog(1, "Time change failed");
+                }
 
                 /* Drop root privileges again */
                 if (sw_uid) swuid(sw_uid);
